src/encrypted_dir.rs

```diff
@@ -0,0 +1,552 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rand::{thread_rng, Rng};
+use std::fs::File;
+use std::io::Error as IoError;
+use std::io::{BufWriter, Cursor, ErrorKind, Read, Write};
+use std::path::{Path, PathBuf};
+
+use crypto::aessafe::AesSafe256Encryptor;
+use crypto::blockmodes::CtrMode;
+use crypto::buffer::{BufferResult, ReadBuffer, RefReadBuffer, RefWriteBuffer, WriteBuffer};
+use crypto::hkdf::hkdf_expand;
+use crypto::hmac::Hmac;
+use crypto::mac::{Mac, MacResult};
+use crypto::pbkdf2::pbkdf2;
+use crypto::sha2::{Sha256, Sha512};
+use crypto::symmetriccipher::{Decryptor, Encryptor};
+
+use tantivy::directory::error::IOError as TvIoError;
+use tantivy::directory::error::{
+    DeleteError, LockError, OpenDirectoryError, OpenReadError, OpenWriteError,
+};
+use tantivy::directory::Directory;
+use tantivy::directory::WatchHandle;
+use tantivy::directory::{
+    AntiCallToken, DirectoryLock, Lock, ReadOnlySource, TerminatingWrite, WatchCallback, WritePtr,
+};
+
+use zeroize::Zeroizing;
+
+use crate::encrypted_stream::{AesReader, AesWriter};
+
+/// KeyBuffer type that makes sure that the buffer is zeroed out before being
+/// dropped.
+type KeyBuffer = Zeroizing<Vec<u8>>;
+
+/// Key derivation result type for our initial key derivation. Consists of a
+/// tuple containing a encryption key, a MAC key, and a random salt.
+type InitialKeyDerivationResult = (KeyBuffer, KeyBuffer, Vec<u8>);
+
+/// Key derivation result for our subsequent key derivations. The salt will be
+/// read from our key file and we will re-derive our encryption and MAC keys.
+type KeyDerivationResult = (KeyBuffer, KeyBuffer);
+
+// The constants here are chosen to be similar to the constants for the Matrix
+// key export format[1].
+// [1] https://matrix.org/docs/spec/client_server/r0.5.0#key-exports
+const KEYFILE: &str = "seshat-index.key";
+// 16 byte random salt.
+const SALT_SIZE: usize = 16;
+// 16 byte random IV for the AES-CTR mode.
+const IV_SIZE: usize = 16;
+// 32 byte or 256 bit encryption keys.
+const KEY_SIZE: usize = 32;
+// 32 byte message authentication code since HMAC-SHA256 is used.
+const MAC_LENGTH: usize = 32;
+// 1 byte for the store version.
+const VERSION: u8 = 1;
+
+#[cfg(test)]
+// Tests don't need to protect against brute force attacks.
+const PBKDF_COUNT: u32 = 10;
+
+#[cfg(not(test))]
+// This is quite a bit lower than the spec since the encrypted key and index
+// will not be public. An attacker would need to get hold of the encrypted index,
+// its key and only then move on to bruteforcing the key. Even if the attacker
+// decrypts the index he wouldn't get to the events itself, only to the index
+// of them.
+const PBKDF_COUNT: u32 = 10_000;
```

Comment:
Shouldn't the number of rounds be configurable? (e.g. save the number of rounds in the key file along with the salt)

Reply:
I was assuming that we'll imply the count from the version. But thinking again if we're going to upstream this it will make more sense to make it configurable and store the rounds.

Reply:
The count is now configurable for the directory and stored in the file for the store_key. Fixed here https://github.com/matrix-org/seshat/pull/15/commits/dc631ea2e568bcff885ca90d4b23d429e73ff4de and here https://github.com/matrix-org/seshat/pull/15/commits/6557025728e076050db1788b8664d8c97ab90add.

Seshat still doesn't expose the setting to users. Let me know if you think that Seshat users should have access to this setting.